### DISCUSSION TITLE
Permissions management changed with new AIR (or Android) ?

### DISCUSSION THREAD

**Discussion by stephanberger4** [airsdk/Adobe-Runtime-Support#1513] [Q&A]
Hi, 

It's been 4 years now that my app is requesting storage access to user in order to read/write a map in the app. 
But it seems now, that use can't access to the map embbed in my app. 
Reason : `Error #3001: File or directory access denied.`

But I didn't chaneg the code... So, is it possible that the last version of AIR, or Android 11, changed the way permission are managed ? 

I'm using, since this 4 years, PermissionCheck ANE from FlashLab. 

My code in my app is this : 

```
Function checkPermissions():void{
		trace("Checking Permissions");
PermissionCheck.init();	
var permissionState:int = PermissionCheck.check(PermissionCheck.SOURCE_STORAGE);
	
		if (permissionState == PermissionCheck.PERMISSION_UNKNOWN || permissionState == PermissionCheck.PERMISSION_DENIED)
		{
			PermissionCheck.request(PermissionCheck.SOURCE_STORAGE, onRequestResult);
			trace("Request Permission Storage");
		}
		else
		{
			trace("permission storage ok");
			requestGeolocationPermission();
		}

		function onRequestResult($state:int):void
		{
			if ($state != PermissionCheck.PERMISSION_GRANTED)
			{
				trace("STORAGE NOT ALLOWED");
				

			}
			else
			{
				//init();
				trace("STORAGE is allowed by user");
			}
		}
	}
```

And in my XML : 

```
  <android>

        <colorDepth>16bit</colorDepth>

        <manifestAdditions><![CDATA[

<manifest android:installLocation="auto">
                 <uses-sdk android:minSdkVersion="19" android:targetSdkVersion="30" />
android:requestLegacyExternalStorage="true"
                 <uses-permission android:name="android.permission.INTERNET"/>
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"
    android:maxSdkVersion="28" />
<uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
```
But when I debug my app : 

The app is asking "Do you authorize Marées NC to access pictures and media on your device ?"
![image](https://user-images.githubusercontent.com/6382334/147979184-dd6ba9c7-8049-43cc-8159-b945785e2f7b.png)

If I click "Authorize", my code is throwing this :
 
```
Checking Permissions
Request Permission Storage
STORAGE NOT ALLOWED
```

So my `function onRequestResult` is returning something else than `PERMISSION_GRANTED`...

And If I want to access the embbed map the app is freezing with this error : 

`Error: Error #3001: File or directory access denied.`

Any idea what could happening (since everything was working before....) ? 

Thank you very much for your help.

**Comment by MalacTheLittle** [comment]
Seems like same problem here #1437

**Reply by stephanberger4** [reply]
Thx. I've just tried with AIR SDK 33.0.2 and it doesn't throw this error.... 
So it seems like a bug..?

Where can I download older AIR SDK versions than the current one 33.1.1 ? (33.1.0 for example)

**Comment by stephanberger4** [comment]
Ok, not that simple... It's, in fact, a bug with the new Adobe AIR **BUT** when using an older version, Google Play refuses the build as it needs the SDK version to be "30". And the older version of Adobe AIR aren't allowing that...
So... **I'M STUCK**

**Comment by FliplineStudios** [comment]
Have you read about the Android 11+ changes with storage permissions?
https://developer.android.com/about/versions/11/privacy/storage

I'm wondering if the errors are related to these changes, and possibly the PermissionCheck ANE doesn't support these new changes if it hasn't been updated in a while.  The newer AIR versions would be built with SDK 30 or 31 and affected by these Android changes/requirements.  That older AIR 33.0.2 you tried was still built targeting 28 or 29 which didn't have these Android requirements yet, so that would likely be why the old version of AIR didn't have this problem (and like you found out, you have to target 30+ now with Google Play....)

**Reply by stephanberger4** [reply]
It's possible that these errors are related to these changes (as Google is saying : 
_If your app targets Android 11, both the `WRITE_EXTERNAL_STORAGE` permission and the `WRITE_MEDIA_STORAGE `privileged permission no longer provide any additional access_.)

. So, I'll have to delete and found a new ANE I guess (or do you think there's a simplier way around ?). 
Is there any tutorial to manage permission (check permission) on Android 11 with ActionScript3 (code that not everybody are familiar with.... unfortunately)
